### ISSUE TITLE
Remove `useOidc` property in providers

### ISSUE THREAD

**Issue by ruben-janssens** [pulumi/pulumi-azure-native#3518]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Currently you are required to set the property `useOidc` on providers to use OIDC for example in GitHub actions. To set/unset this value an apply is needed.

The problem:

For example if you run with `useOidc` set to true in GitHub actions you can no longer preview/refresh your pulumi project locally without first applying which could apply unwanted configurations.

Is it possible to not set this variable and that the provider can detect what to use automatically by looking at set environment variables or configurations?

### Affected area/feature

Provider

**Comment by olafurnielsen** [comment]
From [Pulumi documentation](https://www.pulumi.com/registry/packages/azure-native/installation-configuration/#oidc-pulumi-provider-configuration):

> To use OIDC, either set the Pulumi configuration useOidc via pulumi config set azure-native:useOidc true or set the environment variable ARM_USE_OIDC to “true”.

**Comment by ruben-janssens** [comment]
> From [Pulumi documentation](https://www.pulumi.com/registry/packages/azure-native/installation-configuration/#oidc-pulumi-provider-configuration):
> 
> > To use OIDC, either set the Pulumi configuration useOidc via pulumi config set azure-native:useOidc true or set the environment variable ARM_USE_OIDC to “true”.

This is only the case for the default provider, once I attempt to use a code created provider I must set the `useOidc` property.

**Comment by danielrbradley** [comment]
Hi @ruben-janssens thanks for reaching out.

I don't think we need to completely remove this property - I think we can just tweak the precidence of configuration sources to allow `ARM_USE_OIDC` to be used for explicit providers too where the property is not set either way on the explicit provider constructor.

**Comment by thomas11** [comment]
The fact that the explicit provider didn't pick up the environment variable was a bug which #3525 fixes. This should then unblock you, correct?

I don't think we can easily remove `useOidc` altogether as some variants of OIDC auth use no configuration, so are impossible to detect without `useOidc`.

**Comment by ruben-janssens** [comment]
I will test this when it is released and let you know!
Thanks for picking this up so fast already.

**Comment by ruben-janssens** [comment]
Hi @thomas11 ,
I see now that the property is set to the value of the environment variable but it is stored in state so this prevents me from switching between use OIDC true and false. Can this somehow be resolved to not store it in state?

**Comment by thomas11** [comment]
Hi @ruben-janssens, I think the property is still in your state from before the fix in v2.57.0. After one `pulumi up` with v2.57.0, it should be cleared. Let me know if that's not the case!

**Comment by ruben-janssens** [comment]
@thomas11 Did an apply and sadly it is still in state:
```
{
    "type": "pulumi:providers:azure-native",
    "inputs": {
        "__internal": {},
        "location": "germanywestcentral",
        "subscriptionId": "",
        "useOidc": "true",
        "version": "2.57.1"
    },
    "outputs": {
        "location": "germanywestcentral",
        "subscriptionId": "",
        "useOidc": "true",
        "version": "2.57.1"
    },
    "created": "2024-08-22T09:05:56.107365735Z",
    "modified": "2024-08-22T09:05:56.107365735Z"
},
```

**Comment by ruben-janssens** [comment]
Looks like it was a mistake on my end. Because we are using the automation API some and need a workaround for another issue the config was incorrect for the stack. Seems like it's working now!

**Comment by thomas11** [comment]
Yay, glad to hear that!
